### PR TITLE
Fix build race condition

### LINKED ISSUE #13980: Build race condition
There is a race condition in our build that is causing it to fail with MSB3277 errors. 

``` txt
error MSB3277: Found conflicts between different versions of the same dependent assembly that could not be resolved. These reference conflicts are listed in the build log when log verbosity is set to detailed. [D:\j\workspace\windows_relea---78ad37e5\src\VisualStudio\Core\SolutionExplorerShim\SolutionExplorerShim.csproj] 
```

 In the last 24 hours we've had three failures:

> http://jdash.azurewebsites.net/builds/kind?name=Build&startDate=2016-09-20&viewName=dotnet_roslyn

The actual error from the diagnostic log is the following:

``` txt
There was a conflict between "Microsoft.VisualStudio.Text.Logic, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" and "Microsoft.VisualStudio.Text.Logic, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a". (TaskId:1581)
                        "Microsoft.VisualStudio.Text.Logic, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" was chosen because it was primary and "Microsoft.VisualStudio.Text.Logic, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" was not. (TaskId:1581)
```

The problem appears to be that MSBuild is considering `Binaries\$(Configuration)` when attempting te resolve the `Microsoft.VisualStudio.Text.Logic` reference.  An earlier build component is placing the binary in that directory.  Hence it becomes a race between the component copying it to the output directory and when SolutionExplorerShim builds.

I've labeled this as `urgency-now` because it affects the correctness of the produced DLLs.

### PR DESCRIPTION
**Pull request by jaredpar**
TLDR: artifacts are going to move around in Binaries\Debug and the directory is going to get a lot bigger.

At a high level build projects can be classified into three categories based on how they write output:

- incorrect: a given output path is written to more than once with different contents
- less correct: a given output path is written to more than once but always with the same content
- correct: a given output path is written to exactly once

Today the roslyn build is decidedly “incorrect” as pretty much every file is written directly into Binaries\Debug.  This means it ends up writing pretty much every Visual Studio SDK DLL twice: once for Dev14 and once for Dev15.  For example at various points in the build Binaries\Debug\Microsoft.VisualStudio.Text.Data.dll may refer to Dev14 and at others it’s Dev15.  If this seems like a scary proposition for a build that’s because it is indeed scary and it has real consequences.  By now pretty much everyone on the team has hit the [build race condition](https://github.com/dotnet/roslyn/issues/13980) that is dragging down our PRs. 

The general fix here is to move build outputs into separate directories.  Instead of building to `$(Configuration)` projects now build into say `$(Configuration)\Exes\$(MSBuildProjectFileName)`.  This will have a substantial increase in the size of Binaries.  We will be looking into ways to reduce that.  In the short term though build stability far outweighs the size increase.

This change takes us most of the way to "correct".  There are several places I had to compromise in order to get this initial change in:

- UnitTests still build to a common output folder (one for Dev14, another for Dev15).  Pulling unit tests apart is going to take a bit of work.
- Every project has a `<RoslynProjectType>` entry.  This will go away in the future for most projects.  It's temporarily needed so I can fix roslyn-internal in parallel without taking down the build.
- VSL.Imports.targets is messy.  Unavoidable for now due to the above.  It will get cleaner as I iterate on this.

None of these are relevant to the underlying race condition.  Hence it's okay to push them off.

closes #13980

### CODE REVIEW

**Comment by mmitche** [comment]
@dotnet-bot test internal-perf

**Comment by mmitche** [comment]
@dotnet-bot test internal-perf

**Comment by jaredpar** [comment]
eta failures are known.

File: Program.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace DoNotUse
```
Comment (jasonmalinowski) [inline Program.cs]: What is this and why is it in the root?

File: build/Targets/VSL.Imports.targets [hunk 2]
```diff
@@ -1,17 +1,104 @@
...
+      </PropertyGroup>
+    </When>
+    <When Condition="'$(RoslynProjectType)' == 'Vsix'">
+      <PropertyGroup>
+        <_CopyReferences>false</_CopyReferences>
+        <OutDir>$(OutDir)Vsix\$(MSBuildProjectName)\</OutDir>
+      </PropertyGroup>
+    </When>
+    <When Condition="'$(RoslynProjectType)' == 'Ignore'">
+      <!-- Thinking through these -->
```
Comment (jasonmalinowski) [inline build/Targets/VSL.Imports.targets]: Comment to remove? Seems unactionable.

File: build/Targets/VSL.Imports.targets [hunk 3]
```diff
@@ -1,17 +1,104 @@
...
+    <When Condition="'$(RoslynProjectType)' == 'Dependency'">
+      <!-- Thinking through these -->
+      <PropertyGroup>
+        <OutDir>$(OutDir)Ignore\$(MSBuildProjectName)\</OutDir>
+      </PropertyGroup>
+    </When>
+    <When Condition="'$(RoslynProjectType)' == 'Custom'">
+      <!-- Do nothing -->
+    </When>
+  </Choose>
```
Comment (jasonmalinowski) [inline build/Targets/VSL.Imports.targets]: No `<Otherwise>` to fail the build if one isn't set?

File: build/config/SignToolData.json [hunk 4]
```diff
@@ -3,99 +3,99 @@
...
-            "SDK\\Roslyn.SyntaxVisualizer.Control.dll",
-            "SDK\\Roslyn.SyntaxVisualizer.Extension.dll",
-            "SDK\\Roslyn.Templates.dll",
-            "vbc.exe",
-            "vbccore\\vbc.exe",
-            "VBCSCompiler.exe",
-            "vbi.exe",
-            "vbicore\\vbi.exe"
-          ]
+            "values": [
```
Comment (jasonmalinowski) [inline build/config/SignToolData.json]: Should this be resorted now?

File: build/scripts/create-perftests.ps1 [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# Create the PerfTests directory under Binaries\$(Configuration).  There are still a number
```
Comment (jasonmalinowski) [inline build/scripts/create-perftests.ps1]: Can you document the dependency (or at least explain it here?) I'm curious as I'm looking to delete some of that.

File: src/Compilers/CSharp/CscCore/csc.xml [hunk 6]
```diff
@@ -0,0 +1,244 @@
+<?xml version="1.0"?>
```
Comment (jasonmalinowski) [inline src/Compilers/CSharp/CscCore/csc.xml]: Why is this getting checked in? File building into the wrong place?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/ExpressionCompiler.csproj [hunk 7]
```diff
@@ -12,9 +12,8 @@
     <TargetFrameworkIdentifier>.NETPortable</TargetFrameworkIdentifier>
     <TargetFrameworkVersion>v5.0</TargetFrameworkVersion>
     <ProjectTypeGuids>{786C830F-07A1-408B-BD7F-6EE04809D6DB};{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}</ProjectTypeGuids>
-    <UseCommonOutputDirectory>True</UseCommonOutputDirectory>
-    <!-- Don't transitively copy output files, since everything builds to the same folder. -->
     <CopyNuGetImplementations>false</CopyNuGetImplementations>
```
Comment (jasonmalinowski) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/ExpressionCompiler.csproj]: So what's the policy here for whether this should still be stated?

File: src/Scripting/CSharpTest/CSharpScriptingTest.csproj [hunk 8]
```diff
@@ -18,6 +18,7 @@
     </TargetFrameworkProfile>
     <TargetFrameworkIdentifier>.NETPortable</TargetFrameworkIdentifier>
     <ProjectTypeGuids>{786C830F-07A1-408B-BD7F-6EE04809D6DB};{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}</ProjectTypeGuids>
+    <RoslynProjectType>Ignore</RoslynProjectType>
```
Comment (jasonmalinowski) [inline src/Scripting/CSharpTest/CSharpScriptingTest.csproj]: ...why? What does this mean? Should there be a comment?

File: src/Setup/DevDivInsertionFiles/BuildDevDivInsertionFiles.vb [hunk 9]
```diff
@@ -434,6 +434,15 @@ Public Class BuildDevDivInsertionFiles
         GenerateTestFileDependencyList(NameOf(IntegrationTestFilesExtra), IntegrationTestFilesExtra, insertedFiles)
     End Sub
 
+    Private Function GetPathRelativeToBinaries(p As String) As String
+        Debug.Assert(p.StartsWith(_binDirectory, StringComparison.OrdinalIgnoreCase))
```
Comment (jasonmalinowski) [inline src/Setup/DevDivInsertionFiles/BuildDevDivInsertionFiles.vb]: I'd say no debug assert here. The generate solution code had this same pattern and it was terrifying (partly because the function was actually broken.)

File: src/Setup/DevDivInsertionFiles/BuildDevDivInsertionFiles.vb [hunk 10]
```diff
@@ -434,6 +434,15 @@ Public Class BuildDevDivInsertionFiles
...
+        p = p.Substring(_binDirectory.Length)
+        If Not String.IsNullOrEmpty(p) AndAlso p(0) = "\"c Then
```
Comment (jasonmalinowski) [inline src/Setup/DevDivInsertionFiles/BuildDevDivInsertionFiles.vb]: It's not null, so p.StartsWith("\")?

File: src/Test/DeployDesktopTestRuntime/DeployDesktopTestRuntime.csproj [hunk 11]
```diff
@@ -15,6 +15,7 @@
     <TargetFrameworkIdentifier />
     <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
     <CopyNuGetImplementations>true</CopyNuGetImplementations>
+    <RoslynProjectType>Ignore</RoslynProjectType>
```
Comment (jasonmalinowski) [inline src/Test/DeployDesktopTestRuntime/DeployDesktopTestRuntime.csproj]: Again, why?

File: src/VisualStudio/VisualStudioInteractiveComponents/AssemblyRedirects.cs [hunk 12]
```diff
@@ -5,17 +5,17 @@
 
 [assembly: ProvideRoslynBindingRedirection("Microsoft.CodeAnalysis.Scripting.dll")]
 [assembly: ProvideRoslynBindingRedirection("Microsoft.CodeAnalysis.CSharp.Scripting.dll")]
-[assembly: ProvideRoslynBindingRedirection("Microsoft.CodeAnalysis.VisualBasic.Scripting.dll")]
+// [assembly: ProvideRoslynBindingRedirection("Microsoft.CodeAnalysis.VisualBasic.Scripting.dll")]
```
Comment (jasonmalinowski) [inline src/VisualStudio/VisualStudioInteractiveComponents/AssemblyRedirects.cs]: Delete rather than comment out?

File: Program.cs (hunk 1 above)
Comment (jaredpar) [inline Program.cs]: Mistake.  Will delete it and Program.exe from history.

File: build/config/SignToolData.json (hunk 4 above)
Comment (jaredpar) [inline build/config/SignToolData.json]: But that's so hard and time and oh who am I kidding?

```
:sort
```

File: build/scripts/create-perftests.ps1 (hunk 5 above)
Comment (jaredpar) [inline build/scripts/create-perftests.ps1]: There are both dependencies in DevDivInesrtionFiles.sln and inside roslyn-internal scripts.

File: src/Compilers/CSharp/CscCore/csc.xml (hunk 6 above)
Comment (jaredpar) [inline src/Compilers/CSharp/CscCore/csc.xml]: Good catch.  I messed up the XML file logic.  Will delete from history.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/ExpressionCompiler.csproj (hunk 7 above)
Comment (jaredpar) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/ExpressionCompiler.csproj]: The policy is that projects never include these elements.  I'm working on BuildBoss which will make this policy law.

File: src/Scripting/CSharpTest/CSharpScriptingTest.csproj (hunk 8 above)
Comment (jaredpar) [inline src/Scripting/CSharpTest/CSharpScriptingTest.csproj]: Missed that.  Going to clean up all of the Ignore uses.

File: src/Setup/DevDivInsertionFiles/BuildDevDivInsertionFiles.vb (hunk 9 above)
Comment (jaredpar) [inline src/Setup/DevDivInsertionFiles/BuildDevDivInsertionFiles.vb]: What would you recommend instead?  This invariant should be true here.

File: src/VisualStudio/VisualStudioInteractiveComponents/AssemblyRedirects.cs (hunk 12 above)
Comment (jaredpar) [inline src/VisualStudio/VisualStudioInteractiveComponents/AssemblyRedirects.cs]: That is part of the original change I inheritted.

**Comment by jaredpar** [comment]
@jasonmalinowski responded to feedback.

**Comment by jaredpar** [comment]
I've looked at the individual jobs and they passed.  @jasonmalinowski and I chatted in person and given I removed teh XML files he was fine with signing off.  Github was just down and hence he couldn't change his reply.
